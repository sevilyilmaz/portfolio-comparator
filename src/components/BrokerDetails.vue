<template>
  <div class="accordion-item">
    <h2 class="accordion-header">
      <button :class="headerButtonClasses" type="button" @click="toggleActive">
        {{ name }}
      </button>
    </h2>

    <div :class="contentClasses">
      <div class="accordion-body">
        <div class="row mb-3">
          <label class="col-2 col-form-label">Service Fee:</label>
          <div class="col-sm-6">
            <input
              v-model="internalServiceFee"
              type="text"
              class="form-control"
            />
          </div>

          <div class="col-sm-4 text-end">
            <input
              type="button"
              class="btn btn-danger"
              value="Remove Broker"
              @click="removeBroker"
            />
          </div>
        </div>

        <SecurityList
          v-if="internalSecurities.length"
          :securities="internalSecurities"
        />

        <button type="submit" class="btn btn-primary" @click="addSecurity">
          Add Security
        </button>
      </div>
    </div>
  </div>
</template>

<script lang="ts">
import {
  computed,
  defineComponent,
  onUpdated,
  ref,
  type PropType,
  type SetupContext,
} from 'vue';
import SecurityList from './SecurityList.vue';
import { useId } from '../composables/use-id';
import { useState } from '../composables/use-store';
import { subtractWealthTax } from '../composables/use-wealth-tax';
import { subtractServiceFee } from '../composables/use-service-fee';
import {
  addSecurityValues,
  subtractSecurityExitCost,
} from '../composables/use-securities';
import type { Security } from '../types';
type EmitOption =
  | 'portfolioUpdated'
  | 'portfolioCreated'
  | 'update:serviceFee'
  | 'update:securities'
  | 'removeBroker';

export default defineComponent({
  name: 'BrokerDetails',

  components: { SecurityList },

  props: {
    serviceFee: {
      type: String,
      required: true,
    },
    id: {
      type: String,
      required: true,
    },
    name: {
      type: String,
      required: true,
    },
    securities: {
      type: Array as PropType<Security[]>,
      default: () => [],
    },
  },

  emits: [
    'portfolioUpdated',
    'portfolioCreated',
    'removeBroker',
    'update:serviceFee',
    'update:securities',
  ],

  setup(props, { emit }: SetupContext<EmitOption[]>) {
    const state = useState();
    const active = ref(false);

    const contentClasses = computed(() => ({
      'accordion-collapse': true,
      collapse: true,
      show: active.value,
    }));

    const headerButtonClasses = computed(() => ({
      'accordion-button': true,
      collapsed: !active.value,
    }));

    function toggleActive() {
      active.value = !active.value;
    }

    const internalServiceFee = computed<string>({
      get() {
        return props.serviceFee;
      },
      set(value) {
        emit('update:serviceFee', value);
      },
    });

    const internalSecurities = computed<Security[]>({
      get() {
        return props.securities;
      },
      set(value) {
        emit('update:securities', value);
      },
    });

    const hasServiceFee = computed(() => {
      return internalServiceFee.value;
    });

    const portfolioTable = computed(() => {
      const arr = new Array(state.duration).fill(0).map((_, i) => i);

      const accumulation = arr.reduce((acc, curr) => {
        let currentAccumulation = 0;
        let yearlyDeposit = state.deposit * 12;

        if (curr === 0) {
          currentAccumulation = state.initialDeposit + yearlyDeposit;
        } else {
          currentAccumulation = yearlyDeposit + acc[curr - 1];
        }

        if (hasServiceFee.value) {
          currentAccumulation = subtractServiceFee(
            currentAccumulation,
            internalServiceFee.value
          );
        }

        currentAccumulation = addSecurityValues(
          currentAccumulation,
          internalSecurities.value
        );

        if (state.includeWealthTax) {
          currentAccumulation = subtractWealthTax(
            currentAccumulation,
            state.fiscalStatus
          );
        }

        acc.push(currentAccumulation);

        return acc;
      }, [] as number[]);

      const lastItem = accumulation[accumulation.length - 1];
      const exitValue = subtractSecurityExitCost(
        lastItem,
        internalSecurities.value
      );

      accumulation.push(exitValue);

      return accumulation;
    });

    emit('portfolioCreated', {
      id: props.id,
      name: props.name,
      values: portfolioTable,
    });

    onUpdated(() => {
      emit('portfolioUpdated', {
        id: props.id,
        name: props.name,
        values: portfolioTable,
      });
    });

    function addSecurity() {
      internalSecurities.value.push({
        id: useId(),
        name: '',
        tertd: 0,
        dividend: 0,
        annualReturn: 0,
        fixedTransactionCost: 0,
        variableTransactionCost: 0,
        exitCost: 0,
        allocation: 0,
      });
    }

    function removeBroker() {
      emit('removeBroker', props.id);
    }

    return {
      contentClasses,
      headerButtonClasses,
      toggleActive,
      internalServiceFee,
      internalSecurities,
      portfolioTable,
      addSecurity,
      removeBroker,
    };
  },
});
</script>
