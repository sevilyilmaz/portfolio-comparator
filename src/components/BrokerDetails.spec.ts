import { mount, shallowMount } from '@vue/test-utils';
import BrokerDetails from './BrokerDetails.vue';

export function createWrapper({ props = {}, shallow = true } = {}) {
  const _mount = shallow ? shallowMount : mount;

  return _mount(BrokerDetails, {
    props: {
      id: 'test-id',
      name: 'Test',
      deposit: 1500,
      initialDeposit: 10000,
      duration: 20,
      includeWealthTax: true,
      fiscalStatus: 1,
      serviceFee: '0',
      ...props,
    },
  });
}

describe('BrokerDetails', () => {
  it('should trigger portfolioCreated event on created', () => {
    const wrapper = createWrapper();

    expect(wrapper.emitted('portfolioCreated')).toHaveLength(1);
  });

  describe('portfolioTable', () => {
    it('should show accumulated values without any fees', () => {
      const wrapper = createWrapper({
        props: {
          includeWealthTax: false,
        },
      });

      expect(wrapper.vm.portfolioTable).toMatchSnapshot();
    });

    it('should show accumulated values with wealth tax', () => {
      const wrapper = createWrapper();

      expect(wrapper.vm.portfolioTable).toMatchSnapshot();
    });

    it('should show accumulated values with wealth tax for a couple', () => {
      const wrapper = createWrapper({
        props: {
          fiscalStatus: 2,
        },
      });

      expect(wrapper.vm.portfolioTable).toMatchSnapshot();
    });
  });
});
