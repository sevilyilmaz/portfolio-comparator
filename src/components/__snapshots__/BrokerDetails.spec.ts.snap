// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`BrokerDetails portfolioTable should show accumulated values with wealth tax 1`] = `
Array [
  28000,
  46000,
  63917.6268,
  81729.82986743416,
  99437.22949446016,
  116899.736087601,
  134118.5865799971,
  151097.18072958774,
  167838.8708571497,
  184346.9625081928,
  200624.71510561975,
  216675.34259327952,
  232502.01407054122,
  248107.85441801354,
  263495.9449145335,
  278669.3238455465,
  293630.9871029972,
  308383.88877685036,
  322930.9417383581,
  337275.01821518864,
  337275.01821518864,
]
`;

exports[`BrokerDetails portfolioTable should show accumulated values with wealth tax for a couple 1`] = `
Array [
  28000,
  46000,
  64000,
  82000,
  100000,
  117894.0916,
  135682.8979438439,
  153367.03850892192,
  170947.12912774313,
  188423.7820093813,
  205745.7703366262,
  222826.06322854225,
  239668.03308570807,
  256275.00525325988,
  272650.25867746066,
  288797.0265531082,
  304718.49696191,
  320417.81350195076,
  335898.0759083767,
  351162.3406654195,
  351162.3406654195,
]
`;

exports[`BrokerDetails portfolioTable should show accumulated values without any fees 1`] = `
Array [
  28000,
  46000,
  64000,
  82000,
  100000,
  118000,
  136000,
  154000,
  172000,
  190000,
  208000,
  226000,
  244000,
  262000,
  280000,
  298000,
  316000,
  334000,
  352000,
  370000,
  370000,
]
`;
